lib/Hakyll/Web/Html/Compress.hs

```diff
@@ -0,0 +1,101 @@
+--   This is useful if a page is heavy on code listings.
+--   Don't forget to add @tab-size: n@ to your CSS!
+--
+-- Any of these steps can be disabled, see 'CompressHtmlOpts'.
+
+module Hakyll.Web.Html.Compress
+  ( CompressHtmlOpts(..)
+  , def
+  , compressHtml
+  , compressHtmlCompiler
+  ) where
+
+import           Data.Char
+import           Data.Default
+import qualified Data.Set as S
+import           Hakyll.Core.Compiler
+import           Hakyll.Core.Item
+import           Hakyll.Web.Html
+import           Text.HTML.TagSoup
+
+-- | The configuration for the HTML compression.
+data CompressHtmlOpts = CompressHtmlOpts
+  { choRemoveComments :: Bool          -- ^ Whether to remove comments.
+  , choCompressWhitespace :: Bool      -- ^ Whether to remove excessive whitespaces.
+  , choTabSize :: Maybe Int            -- ^ Replace this many spaces in <pre> with @\t@ (if 'choCompressWhitespace' is set).
+  }
+
+instance Default CompressHtmlOpts where
+  def = CompressHtmlOpts
+          { choRemoveComments = True
+          , choCompressWhitespace = True
+          , choTabSize = Nothing
+          }
+
+-- | Compiler form of 'compressHtml'.
+compressHtmlCompiler :: CompressHtmlOpts -> Item String -> Compiler (Item String)
+compressHtmlCompiler opts item = pure $ compressHtml opts <$> item
+
+-- | Compresses an HTML string according to the given configuration.
+compressHtml :: CompressHtmlOpts -> String -> String
+compressHtml CompressHtmlOpts{ .. } = withTagList go
+  where
+    go = foldr (.) id
+        $ [ f
+          | (True, f) <- [ (choRemoveComments, removeComments)
+                         , (choCompressWhitespace, compressWS choTabSize)
+                         ]
+          ]
+
+removeComments :: [Tag String] -> [Tag String]
+removeComments = filter (not . isTagComment)
+
+compressWS :: Maybe Int -> [Tag String] -> [Tag String]
+compressWS maybeTabSize = go mempty
+  where
+    go stack =
+      \case [] -> []
+            (tag@(TagClose n) : rest) -> tag : go (S.delete n stack) rest
+            (tag@(TagOpen n _) : rest) -> tag : go (S.insert n stack) rest
+            (tag@(TagText text) : rest)
+              -- all spaces within a <pre> are important, but we can replace them with tabs
+              | "pre" `S.member` stack -> case maybeTabSize of
+                                               Nothing -> tag : go stack rest
+                                               Just tabSize -> TagText (collapseIntoTabs (tabSize - 1) text) : go stack rest
+              | otherwise -> let text' = collapseSpaces text
+                              in case text' of
+                                      [] -> go stack rest
+                                      _  -> TagText text' : go stack rest
+            (tag : rest) -> tag : go stack rest
+
+collapseSpaces :: String -> String
+collapseSpaces = go
+  where
+    go [] = []
+    go [c] = [c]
+    go (c1 : c2 : rest)
+      | isSpace c1 && isSpace c2 = go (c2 : rest)
```

Comment:
This doesn't make a distinction between various kinds of spaces, most notably no-break spaces. I can't find the reference in the spec, but browsers definitely avoid collapsing consecutive no-break spaces. Authors could rely on this behaviour, so can you please make it so no-breaking spaces are preserved?

Reply:
Sure, thanks for catching this! I was considering non-breaking spaces only as `&nbsp;`, and didn't think about `isSpace` swallowing those as well.

I can add a check against `c1` not being one of U+00A0, U+2007, U+202F, or U+2060. I think that should do.

Reply:
Apparently [there are dozens of kinds of spaces in Unicode](https://en.wikipedia.org/wiki/Template:Whitespace_(Unicode)), so I think it's better to be restrictive here: check that `c1` and `c2` are one of space, \t, \v, \n, \r, or \f. This seems like a safe subset that can be expanded later.